src/core/vector/qgsvectorlayerprofilegenerator.cpp

```diff
@@ -802,423 +814,451 @@ bool QgsVectorLayerProfileGenerator::generateProfileForPoints()
+  const double *inY = intersectionLine->yData();
+  const double *inZ = intersectionLine->is3D() ? intersectionLine->zData() : nullptr;
+  double *outX = newX.data();
+  double *outY = newY.data();
+  double *outZ = newZ.data();
+  double *outDistance = newDistance.data();
+
+  double lastDistanceAlongProfileCurve = 0.0;
+  QVector< double > extrudedZ;
+  double *extZOut = nullptr;
+  double extrusion = 0;
+
+  if ( mExtrusionEnabled )
+  {
+    extrudedZ.resize( numPoints );
+    extZOut = extrudedZ.data();
 
-          mResults->mDistanceToHeightMap.insert( distance, z );
-        }
+    extrusion = mDataDefinedProperties.valueAsDouble( QgsMapLayerElevationProperties::ExtrusionHeight, mExpressionContext, mExtrusionHeight );
+  }
 
-        QgsVectorLayerProfileResults::Feature resultFeature;
-        resultFeature.featureId = feature.id();
+  QString lastError;
+  for ( int i = 0 ; ! mFeedback->isCanceled() && i < numPoints; ++i )
+  {
+    double x = *inX++;
+    double y = *inY++;
+    double z = inZ ? *inZ++ : 0;
 
-        if ( mExtrusionEnabled )
-        {
-          std::unique_ptr< QgsLineString > ring = std::make_unique< QgsLineString >( newX, newY, newZ );
-          std::unique_ptr< QgsLineString > extrudedRing = std::make_unique< QgsLineString >( newX, newY, extrudedZ );
-          std::unique_ptr< QgsLineString > reversedExtrusion( extrudedRing->reversed() );
-          ring->append( reversedExtrusion.get() );
-          ring->close();
-          resultFeature.geometry = QgsGeometry( new QgsPolygon( ring.release() ) );
-
-
-          std::unique_ptr< QgsLineString > distanceVHeightRing = std::make_unique< QgsLineString >( newDistance, newZ );
-          std::unique_ptr< QgsLineString > extrudedDistanceVHeightRing = std::make_unique< QgsLineString >( newDistance, extrudedZ );
-          std::unique_ptr< QgsLineString > reversedDistanceVHeightExtrusion( extrudedDistanceVHeightRing->reversed() );
-          distanceVHeightRing->append( reversedDistanceVHeightExtrusion.get() );
-          distanceVHeightRing->close();
-          resultFeature.crossSectionGeometry = QgsGeometry( new QgsPolygon( distanceVHeightRing.release() ) );
-        }
-        else
-        {
-          resultFeature.geometry = QgsGeometry( new QgsLineString( newX, newY, newZ ) );
-          resultFeature.crossSectionGeometry = QgsGeometry( new QgsLineString( newDistance, newZ ) );
-        }
-        mResults->features[resultFeature.featureId].append( resultFeature );
+    QgsPoint interpolatedPoint( x, y, z ); // general case (not a triangle)
+
+    *outX++ = x;
+    *outY++ = y;
+    if ( triangle->exteriorRing()->numPoints() <= 4 ) // triangle case
+    {
+      QgsPoint tmpPt = interpolatePointOnTriangle( triangle, x, y );
+      if ( ! tmpPt.isEmpty() ) // point x,y inside the triangle
+      {
+        interpolatedPoint = tmpPt;
       }
     }
-  };
+    *outZ++ = std::isnan( interpolatedPoint.z() ) ? 0.0 : interpolatedPoint.z();
```

Comment:
There's a lot of duplicate isnan checks here, they all are non-free and so should only be checked once.